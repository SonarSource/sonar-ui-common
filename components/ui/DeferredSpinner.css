/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
.deferred-spinner {
  position: relative;
  vertical-align: middle;
  width: 16px;
  height: 16px;
  border: 2px solid var(--blue);
  border-radius: 50%;
  animation: spin 0.75s infinite linear;
}

.deferred-spinner:before,
.deferred-spinner:after {
  left: -2px;
  top: -2px;
  display: none;
  position: absolute;
  content: '';
  width: inherit;
  height: inherit;
  border: inherit;
  border-radius: inherit;
}

.deferred-spinner,
.deferred-spinner:before,
.deferred-spinner:after {
  display: inline-block;
  box-sizing: border-box;
  border-color: transparent;
  border-top-color: var(--blue);
  animation-duration: 1.2s;
}

.deferred-spinner:before {
  transform: rotate(120deg);
}

.deferred-spinner:after {
  transform: rotate(240deg);
}

.deferred-spinner-placeholder {
  position: relative;
  display: inline-block;
  vertical-align: middle;
  width: 16px;
  height: 16px;
  visibility: hidden;
}

@keyframes spin {
  from {
    transform: rotate(0deg);
  }

  to {
    transform: rotate(360deg);
  }
}
