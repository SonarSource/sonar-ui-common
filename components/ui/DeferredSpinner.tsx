/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import * as classNames from 'classnames';
import './DeferredSpinner.css';

interface Props {
  children?: React.ReactNode;
  className?: string;
  loading?: boolean;
  customSpinner?: JSX.Element;
  timeout?: number;
}

interface State {
  showSpinner: boolean;
}

export default class DeferredSpinner extends React.PureComponent<Props, State> {
  timer: any;

  static defaultProps = {
    timeout: 100
  };

  state: State = { showSpinner: false };

  componentDidMount() {
    if (this.props.loading == null || this.props.loading === true) {
      this.startTimer();
    }
  }

  componentWillReceiveProps(nextProps: Props) {
    if (this.props.loading === false && nextProps.loading === true) {
      this.stopTimer();
      this.startTimer();
    }
    if (this.props.loading === true && nextProps.loading === false) {
      this.stopTimer();
      this.setState({ showSpinner: false });
    }
  }

  componentWillUnmount() {
    this.stopTimer();
  }

  startTimer = () => {
    this.timer = setTimeout(() => this.setState({ showSpinner: true }), this.props.timeout);
  };

  stopTimer = () => {
    clearTimeout(this.timer);
  };

  render() {
    if (this.state.showSpinner) {
      return (
        this.props.customSpinner || (
          <i className={classNames('deferred-spinner', this.props.className)} />
        )
      );
    }
    return this.props.children || null;
  }
}
