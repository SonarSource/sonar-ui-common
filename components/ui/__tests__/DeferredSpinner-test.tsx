/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { mount } from 'enzyme';
import * as React from 'react';
import DeferredSpinner from '../DeferredSpinner';

jest.useFakeTimers();

it('renders spinner after timeout', () => {
  const spinner = mount(<DeferredSpinner />);
  expect(spinner).toMatchSnapshot();
  jest.runAllTimers();
  spinner.update();
  expect(spinner).toMatchSnapshot();
});

it('add custom className', () => {
  const spinner = mount(<DeferredSpinner className="foo" />);
  jest.runAllTimers();
  spinner.update();
  expect(spinner).toMatchSnapshot();
});

it('renders children before timeout', () => {
  const spinner = mount(
    <DeferredSpinner>
      <div>foo</div>
    </DeferredSpinner>
  );
  expect(spinner).toMatchSnapshot();
  jest.runAllTimers();
  spinner.update();
  expect(spinner).toMatchSnapshot();
});

it('is controlled by loading prop', () => {
  const spinner = mount(
    <DeferredSpinner loading={false}>
      <div>foo</div>
    </DeferredSpinner>
  );
  expect(spinner).toMatchSnapshot();
  spinner.setProps({ loading: true });
  expect(spinner).toMatchSnapshot();
  jest.runAllTimers();
  spinner.update();
  expect(spinner).toMatchSnapshot();
  spinner.setProps({ loading: false });
  expect(spinner).toMatchSnapshot();
});
