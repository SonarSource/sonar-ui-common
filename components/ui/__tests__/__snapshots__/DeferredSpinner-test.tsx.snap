// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add custom className 1`] = `
<DeferredSpinner
  className="foo"
  timeout={100}
>
  <i
    className="deferred-spinner foo"
  />
</DeferredSpinner>
`;

exports[`is controlled by loading prop 1`] = `
<DeferredSpinner
  loading={false}
  timeout={100}
>
  <div>
    foo
  </div>
</DeferredSpinner>
`;

exports[`is controlled by loading prop 2`] = `
<DeferredSpinner
  loading={true}
  timeout={100}
>
  <div>
    foo
  </div>
</DeferredSpinner>
`;

exports[`is controlled by loading prop 3`] = `
<DeferredSpinner
  loading={true}
  timeout={100}
>
  <i
    className="deferred-spinner"
  />
</DeferredSpinner>
`;

exports[`is controlled by loading prop 4`] = `
<DeferredSpinner
  loading={false}
  timeout={100}
>
  <div>
    foo
  </div>
</DeferredSpinner>
`;

exports[`renders children before timeout 1`] = `
<DeferredSpinner
  timeout={100}
>
  <div>
    foo
  </div>
</DeferredSpinner>
`;

exports[`renders children before timeout 2`] = `
<DeferredSpinner
  timeout={100}
>
  <i
    className="deferred-spinner"
  />
</DeferredSpinner>
`;

exports[`renders spinner after timeout 1`] = `
<DeferredSpinner
  timeout={100}
/>
`;

exports[`renders spinner after timeout 2`] = `
<DeferredSpinner
  timeout={100}
>
  <i
    className="deferred-spinner"
  />
</DeferredSpinner>
`;
